### PR TITLE
 Add meal plan / recipe interfaces. #90 

### LINKED ISSUE #90: implement meal plan and recipes interface
see https://github.com/custom-components/grocy/issues/45

### PR DESCRIPTION
**Pull request by nervetattoo**
##  Add meal plan / recipe interfaces

This adds `grocy.meal_plan(get_details)` and `grocy.recipe(id)`.

Fixes #90

## Checklist
  - [x] The code change is tested and works locally.
  - [x] tests pass. **Your PR won't be merged unless tests pass**
  - [x] There is no commented out code in this PR

### CODE REVIEW

File: pygrocy/grocy_api_client.py [hunk 1]
```diff
@@ -40,6 +40,82 @@ def note(self) -> str:
...
+    def __init__(self, parsed_json):
+        self._id = parse_int(parsed_json.get('id'))
+        self._day = parse_date(parsed_json.get('day'))
+        self._type = parsed_json.get('type')
+        self._recipe_id = parse_int(parsed_json.get('recipe_id'))
+        self._recipe_servings = parse_int(parsed_json.get('recipe_servings'))
+        self._note = parsed_json.get('note', None)
+        self._product_id = parsed_json.get('product_id')
+        self._product_amount = parse_float(parsed_json.get('product_amount'), 0)
+        self._product_qu_id = parsed_json.get('product_amount')
```
Comment (SebRut) [inline pygrocy/grocy_api_client.py:53]: That doesn't seem right,  maybe a c&p error? :)

File: pygrocy/grocy_api_client.py [hunk 2]
```diff
@@ -40,6 +40,82 @@ def note(self) -> str:
...
+        self._desired_servings = parse_int(parsed_json.get('desired_servings'))
+        self._picture_file_name = parsed_json.get('picture_file_name')
+        self._row_created_timestamp = parse_date(parsed_json.get('row_created_timestamp'))
+        self._userfields = parsed_json.get('userfields')
+
+    @property
+    def id(self) -> int:
+        return self._id
+
+    def picture(self, width: int = 400):
```
Comment (SebRut) [inline pygrocy/grocy_api_client.py:93]: I'd like to have a more meaningful method name, maybe something like `get_full_picture_url`?

File: pygrocy/grocy.py [hunk 3]
```diff
@@ -324,6 +326,42 @@ def id(self) -> str:
...
+    @property
+    def recipe_servings(self) -> int:
+        return self._recipe_servings
+
+    @property
+    def note(self) -> str:
+        return self._note
+
+    @property
+    def recipe(self) -> RecipeDetailsResponse:
```
Comment (SebRut) [inline pygrocy/grocy.py:358]: I'd like to keep "response classes" out of the publicly available API. Could you add an extra class `RecipeDetails` similar to `MealPlanItem` and `MealPlanResponse`?

File: pygrocy/grocy.py [hunk 4]
```diff
@@ -324,6 +326,42 @@ def id(self) -> str:
     def name(self) -> str:
         return self._name
 
+class MealPlanItem(object):
```
Comment (SebRut) [inline pygrocy/grocy.py:329]: Does this represent a meal plan or a meal plan's item?

File: pygrocy/grocy.py [hunk 5]
```diff
@@ -434,3 +472,15 @@ def tasks(self) -> List[Task]:
...
+    def meal_plan(self, get_details: bool = False) -> List[MealPlanItem]:
+        raw_meal_plan = self._api_client.get_meal_plan()
+        meal_plan = [MealPlanItem(data) for data in raw_meal_plan]
+
+        if get_details:
+            for item in meal_plan:
+                item.get_details(self._api_client)
+        return meal_plan
+
+    def recipe(self, recipe_id: int) -> RecipeDetailsResponse:
```
Comment (SebRut) [inline pygrocy/grocy.py:485]: Same as with the `recipe` property of `MealPlanItem` please map the response to a `RecipeDetails` object.

**Comment by SebRut** [comment]
I'm fine with coverage dropping a bit, if the code looks good to me otherwise :)

File: pygrocy/grocy.py (hunk 4 above)
Comment (nervetattoo) [reply pygrocy/grocy.py:329]: A meal plan item. A meal plan is just a list of meal plan items, which apparently are stored as recipes in the underlying data model with day properties.

**Comment by nervetattoo** [comment]
Fixed. Codefactor is not telling me where that complexity is and I can't really see it with my bare eyes 🙄

**Comment by SebRut** [comment]
> Fixed. Codefactor is not telling me where that complexity is and I can't really see it with my bare eyes 🙄

I guess `grocy.py` is getting to big and complex, I'll suppress the warning.

**Comment by nervetattoo** [comment]
> I guess `grocy.py` is getting to big and complex, I'll suppress the warning.

Alright, so probably just by adding another class to the file then. If I end up doing more work on this I'll look into splitting it up I guess

**Comment by SebRut** [comment]
> Alright, so probably just by adding another class to the file then. If I end up doing more work on this I'll look into splitting it up I guess

I created #110 to fix this. Not an issue you should have to fix with this.

**Comment by SebRut** [comment]
LGTM. I'd merge this PR if you're done with this.

**Comment by nervetattoo** [comment]
I'm done yeah. LGTM

**Comment by SebRut** [comment]
Merged. I'll try to create a new release this afternoon.

**Pull request merged**